File: cwt-spring/src/main/java/com/cwtsite/cwt/domain/game/service/GameService.java
package com.cwtsite.cwt.domain.game.service;

import com.cwtsite.cwt.core.FileValidator;
import com.cwtsite.cwt.domain.configuration.entity.Configuration;
import com.cwtsite.cwt.domain.configuration.entity.enumeratuion.ConfigurationKey;
import com.cwtsite.cwt.domain.configuration.service.ConfigurationService;
import com.cwtsite.cwt.domain.game.entity.Game;
import com.cwtsite.cwt.domain.game.entity.Rating;
import com.cwtsite.cwt.domain.game.entity.Replay;
import com.cwtsite.cwt.domain.game.entity.enumeration.RatingType;
import com.cwtsite.cwt.domain.group.entity.Group;
import com.cwtsite.cwt.domain.group.service.GroupRepository;
import com.cwtsite.cwt.domain.group.service.GroupService;
import com.cwtsite.cwt.domain.playoffs.service.PlayoffService;
import com.cwtsite.cwt.domain.tournament.entity.Tournament;
import com.cwtsite.cwt.domain.tournament.entity.enumeration.TournamentStatus;
import com.cwtsite.cwt.domain.tournament.exception.IllegalTournamentStatusException;
import com.cwtsite.cwt.domain.tournament.service.TournamentService;
import com.cwtsite.cwt.domain.user.repository.UserRepository;
import com.cwtsite.cwt.domain.user.repository.entity.User;
import com.cwtsite.cwt.domain.user.service.UserService;
import com.cwtsite.cwt.entity.Comment;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.StringUtils;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Component
public class GameService {

    private final GameRepository gameRepository;
    private final RatingRepository ratingRepository;
    private final TournamentService tournamentService;
    private final GroupRepository groupRepository;
    private final UserRepository userRepository;
    private final GroupService groupService;
    private final CommentRepository commentRepository;
    private final ConfigurationService configurationService;
    private final UserService userService;
    private final PlayoffService playoffService;

    @Autowired
    public GameService(GameRepository gameRepository, TournamentService tournamentService, GroupRepository groupRepository,
                       UserRepository userRepository, GroupService groupService, RatingRepository ratingRepository,
                       CommentRepository commentRepository, ConfigurationService configurationService, UserService userService,
                       PlayoffService playoffService) {
        this.gameRepository = gameRepository;
        this.tournamentService = tournamentService;
        this.groupRepository = groupRepository;
        this.userRepository = userRepository;
        this.groupService = groupService;
        this.ratingRepository = ratingRepository;
        this.commentRepository = commentRepository;
        this.configurationService = configurationService;
        this.userService = userService;
        this.playoffService = playoffService;
    }

    @Transactional
    public Game reportGame(long homeUser, long awayUser, int scoreHome, int scoreAway, MultipartFile replay)
            throws InvalidOpponentException, InvalidScoreException, IllegalTournamentStatusException,
            FileValidator.UploadSecurityException, FileValidator.IllegalFileContentTypeException,
            FileValidator.FileEmptyException, FileValidator.FileTooLargeException, FileValidator.IllegalFileExtension,
            IOException {
        FileValidator.validate(replay, 150000, Arrays.asList("application/x-rar", "application/zip"), Arrays.asList("rar", "zip"));
        final Game reportedGame = reportGame(homeUser, awayUser, scoreHome, scoreAway);
        reportedGame.setReplay(new Replay(replay.getBytes(), replay.getContentType(), StringUtils.getFilenameExtension(replay.getOriginalFilename())));
        return gameRepository.save(reportedGame);
    }

    public String createReplayFileName(Game game) {
        return String.format(
                "%s_%s_%s-%s_%s.%s",
                game.getId(),
                game.getHomeUser().getUsername().replaceAll("[^a-zA-Z0-9-_\\\\.]", "_"),
                game.getScoreHome(), game.getScoreAway(),
                game.getAwayUser().getUsername().replaceAll("[^a-zA-Z0-9-_\\\\.]", "_"),
                game.getReplay().getExtension());
    }

    @Transactional
    public Game reportGame(long homeUserId, long awayUserId, int homeScore, int awayScore)
            throws InvalidOpponentException, InvalidScoreException, IllegalTournamentStatusException {
        final Tournament currentTournament = tournamentService.getCurrentTournament();
        final Configuration bestOfValue = getBestOfValue(currentTournament.getStatus());

        if (homeScore + awayScore != Integer.valueOf(bestOfValue.getValue())) {
            throw new InvalidScoreException(String.format(
                    "Score %s-%s should have been best of %s.",
                    homeScore, awayScore, bestOfValue.getValue()));
        } else if (homeScore < 0 || awayScore < 0) {
            throw new InvalidScoreException(String.format(
                    "Score %s-%s should not include negative scores.", homeScore, awayScore));
        }

        final User homeUser = userRepository.findById(homeUserId).orElseThrow(RuntimeException::new);
        final List<User> remainingOpponents = userService.getRemainingOpponents(homeUser);
        final User awayUser = userRepository.findById(awayUserId).orElseThrow(RuntimeException::new);

        if (!remainingOpponents.contains(awayUser)) {
            throw new InvalidOpponentException(String.format(
                    "Opponent %s is not in %s",
                    awayUser.getId(), remainingOpponents.stream().map(User::getId).collect(Collectors.toList())));
        }

        Game reportedGame;

        if (currentTournament.getStatus() == TournamentStatus.GROUP) {
            final Group group = groupRepository.findByTournamentAndUser(currentTournament, awayUser);

            final Game game = new Game();

            game.setScoreHome(homeScore);
            game.setScoreAway(awayScore);
            game.setTournament(currentTournament);
            game.setHomeUser(homeUser);
            game.setAwayUser(awayUser);
            game.setReporter(homeUser);

            game.setGroup(group);

            groupService.calcTableByGame(game);
            reportedGame = gameRepository.save(game);
        } else if (currentTournament.getStatus() == TournamentStatus.PLAYOFFS) {
            final Game playoffGameToBeReported =
                    gameRepository.findNextPlayoffGameForUser(currentTournament, homeUser);

            if (!Arrays.asList(playoffGameToBeReported.getHomeUser(), playoffGameToBeReported.getAwayUser())
                    .containsAll(Arrays.asList(homeUser, awayUser))) {
                throw new InvalidOpponentException(String.format(
                        "Next playoff game is expected to be %s vs. %s.",
                        homeUser.getUsername(), awayUser.getUsername()));
            }

            playoffGameToBeReported.setReporter(homeUser);

            if (playoffGameToBeReported.getHomeUser().equals(homeUser)) {
                playoffGameToBeReported.setScoreHome(homeScore);
                playoffGameToBeReported.setScoreAway(awayScore);
            } else {
                playoffGameToBeReported.setScoreHome(awayScore);
                playoffGameToBeReported.setScoreAway(homeScore);
            }

            reportedGame = gameRepository.save(playoffGameToBeReported);
            playoffService.advanceByGame(reportedGame);
        } else {
            throw new IllegalTournamentStatusException(TournamentStatus.GROUP, TournamentStatus.PLAYOFFS);
        }

        return reportedGame;
    }

    public Configuration getBestOfValue(TournamentStatus tournamentStatus) {
        ConfigurationKey configurationKey;

        if (tournamentStatus == TournamentStatus.GROUP) {
            configurationKey = ConfigurationKey.GROUP_GAMES_BEST_OF;
        } else if (tournamentStatus == TournamentStatus.PLAYOFFS) {
            configurationKey = playoffService.onlyFinalGamesAreLeftToPlay()
                    ? ConfigurationKey.FINAL_GAME_BEST_OF
                    : ConfigurationKey.PLAYOFF_GAMES_BEST_OF;
        } else {
            throw new IllegalTournamentStatusException(TournamentStatus.GROUP, TournamentStatus.PLAYOFFS);
        }

        return configurationService.getOne(configurationKey);
    }

    public List<Game> saveAll(final List<Game> games) {
        return gameRepository.saveAll(games);
    }

    public Optional<Game> get(long id) {
        return gameRepository.findById(id);
    }

    public Rating rateGame(long gameId, Long userId, RatingType type) {
        final User user = userRepository.findById(userId)
                .orElseThrow(IllegalArgumentException::new);
        final Game game = gameRepository.findById(gameId)
                .orElseThrow(IllegalArgumentException::new);
        return ratingRepository.save(new Rating(type, user, game));
    }

    public Comment commentGame(long gameId, Long userId, String body) {
        final User user = userRepository.findById(userId)
                .orElseThrow(IllegalArgumentException::new);
        final Game game = gameRepository.findById(gameId)
                .orElseThrow(IllegalArgumentException::new);
        return commentRepository.save(new Comment(body, user, game));
    }

    public class InvalidScoreException extends RuntimeException {
        InvalidScoreException(String message) {
            super(message);
        }
    }

    public class InvalidOpponentException extends RuntimeException {
        InvalidOpponentException(String message) {
            super(message);
        }
    }
}

File: cwt-spring/src/main/java/com/cwtsite/cwt/domain/game/service/GameService.kt
package com.cwtsite.cwt.domain.game.service

import com.cwtsite.cwt.controller.RestException
import com.cwtsite.cwt.core.news.PublishNews
import com.cwtsite.cwt.domain.bet.entity.Bet
import com.cwtsite.cwt.domain.bet.service.BetRepository
import com.cwtsite.cwt.domain.configuration.entity.Configuration
import com.cwtsite.cwt.domain.configuration.entity.enumeratuion.ConfigurationKey
import com.cwtsite.cwt.domain.configuration.service.ConfigurationService
import com.cwtsite.cwt.domain.game.entity.Game
import com.cwtsite.cwt.domain.game.entity.GameStats
import com.cwtsite.cwt.domain.game.entity.Rating
import com.cwtsite.cwt.domain.game.entity.enumeration.RatingType
import com.cwtsite.cwt.domain.group.service.GroupRepository
import com.cwtsite.cwt.domain.group.service.GroupService
import com.cwtsite.cwt.domain.message.service.MessageNewsType
import com.cwtsite.cwt.domain.playoffs.service.PlayoffService
import com.cwtsite.cwt.domain.playoffs.service.TreeService
import com.cwtsite.cwt.domain.schedule.service.ScheduleService
import com.cwtsite.cwt.domain.tournament.entity.Tournament
import com.cwtsite.cwt.domain.tournament.entity.enumeration.TournamentStatus
import com.cwtsite.cwt.domain.tournament.exception.IllegalTournamentStatusException
import com.cwtsite.cwt.domain.tournament.service.TournamentService
import com.cwtsite.cwt.domain.user.repository.UserRepository
import com.cwtsite.cwt.domain.user.repository.entity.User
import com.cwtsite.cwt.domain.user.service.UserService
import com.cwtsite.cwt.entity.Comment
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import org.json.JSONArray
import org.json.JSONObject
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import java.sql.Timestamp
import java.text.SimpleDateFormat
import java.util.*
import kotlin.math.ceil

@Component
class GameService

@Autowired
constructor(private val gameRepository: GameRepository,
            private val tournamentService: TournamentService,
            private val groupRepository: GroupRepository,
            private val userRepository: UserRepository,
            private val groupService: GroupService,
            private val ratingRepository: RatingRepository,
            private val commentRepository: CommentRepository,
            private val configurationService: ConfigurationService,
            private val userService: UserService,
            private val playoffService: PlayoffService,
            private val betRepository: BetRepository,
            private val scheduleService: ScheduleService,
            private val treeService: TreeService,
            private val gameStatsRepository: GameStatsRepository) {

    fun createReplayFileName(game: Game): String {
        return String.format(
                "%s_%s_%s-%s_%s.%s",
                game.id,
                game.homeUser!!.username.replace("[^a-zA-Z0-9-_\\\\.]".toRegex(), "_"),
                game.scoreHome, game.scoreAway,
                game.awayUser!!.username.replace("[^a-zA-Z0-9-_\\\\.]".toRegex(), "_"),
                game.replay!!.extension)
    }

    @Transactional
    @Throws(InvalidOpponentException::class, InvalidScoreException::class, IllegalTournamentStatusException::class)
    @PublishNews
    fun reportGame(homeUserId: Long, awayUserId: Long, homeScore: Int, awayScore: Int, persist: Boolean = true): Game {
        val currentTournament = tournamentService.getCurrentTournament()
                ?: throw RestException("There's no tournament currently.", HttpStatus.BAD_REQUEST, null)
        val bestOfValue = Integer.valueOf(getBestOfValue(currentTournament.status).value)
        val winnerScore = ceil(bestOfValue.toDouble() / 2)

        if (homeScore.toDouble() != winnerScore && awayScore.toDouble() != winnerScore || homeScore + awayScore > bestOfValue) {
            throw InvalidScoreException(String.format(
                    "Score %s-%s should have been best of %s.",
                    homeScore, awayScore, bestOfValue))
        } else if (homeScore < 0 || awayScore < 0) {
            throw InvalidScoreException(String.format(
                    "Score %s-%s should not include negative scores.", homeScore, awayScore))
        }

        val homeUser = userRepository.findById(homeUserId).orElseThrow<RuntimeException> { throw RuntimeException() }
        val remainingOpponents = userService.getRemainingOpponents(homeUser)
        val awayUser = userRepository.findById(awayUserId).orElseThrow<RuntimeException> { throw RuntimeException() }

        if (!remainingOpponents.contains(awayUser)) {
            throw InvalidOpponentException(String.format(
                    "Opponent %s is not in %s",
                    awayUser.id, remainingOpponents.map { it.id }))
        }

        val reportedGame: Game

        if (currentTournament.status == TournamentStatus.GROUP) {
            val group = groupRepository.findByTournamentAndUser(currentTournament, awayUser)

            val game = Game(tournament = currentTournament)

            game.scoreHome = homeScore
            game.scoreAway = awayScore
            game.tournament = currentTournament
            game.homeUser = homeUser
            game.awayUser = awayUser
            game.reporter = homeUser

            game.group = group

            groupService.calcTableByGame(game)
            reportedGame = if (persist) gameRepository.save(game) else game
        } else if (currentTournament.status == TournamentStatus.PLAYOFFS) {
            val playoffGameToBeReported = gameRepository.findNextPlayoffGameForUser(currentTournament, homeUser)

            if (!listOf(playoffGameToBeReported.homeUser, playoffGameToBeReported.awayUser)
                            .containsAll(listOf(homeUser, awayUser))) {
                throw InvalidOpponentException(String.format(
                        "Next playoff game is expected to be %s vs. %s.",
                        homeUser.username, awayUser.username))
            }

            playoffGameToBeReported.reporter = homeUser

            if (playoffGameToBeReported.homeUser == homeUser) {
                playoffGameToBeReported.scoreHome = homeScore
                playoffGameToBeReported.scoreAway = awayScore
            } else {
                playoffGameToBeReported.scoreHome = awayScore
                playoffGameToBeReported.scoreAway = homeScore
            }

            reportedGame = gameRepository.save(playoffGameToBeReported)
            playoffService.advanceByGame(reportedGame)
        } else {
            throw IllegalTournamentStatusException(TournamentStatus.GROUP, TournamentStatus.PLAYOFFS)
        }

        GlobalScope.launch {
            scheduleService.delete(
                    scheduleService.findByPairing(reportedGame.homeUser!!, reportedGame.awayUser!!) ?: return@launch)
        }
        reportedGame.reportedAt = Timestamp(System.currentTimeMillis())
        return reportedGame
    }

    fun getBestOfValue(tournamentStatus: TournamentStatus): Configuration {
        val configurationKey = if (tournamentStatus == TournamentStatus.GROUP) {
            ConfigurationKey.GROUP_GAMES_BEST_OF
        } else if (tournamentStatus == TournamentStatus.PLAYOFFS) {
            if (treeService.onlyFinalGamesAreLeftToPlay())
                ConfigurationKey.FINAL_GAME_BEST_OF
            else
                ConfigurationKey.PLAYOFF_GAMES_BEST_OF
        } else {
            throw IllegalTournamentStatusException(TournamentStatus.GROUP, TournamentStatus.PLAYOFFS)
        }
        return configurationService.getOne(configurationKey)
    }

    @Transactional
    @Throws(PlayoffService.PlayoffGameNotVoidableException::class, IllegalStateException::class)
    @PublishNews
    fun voidGame(game: Game): Game {
        val currentTournament = tournamentService.getCurrentTournament()

        if (currentTournament == null || game.tournament != currentTournament)
            throw IllegalStateException("Can only void games of current tournament.")

        game.voided = true

        return when (game.tournament.status) {
            TournamentStatus.GROUP -> groupService.reverseStandingsByGame(game)
            TournamentStatus.PLAYOFFS -> playoffService.voidPlayoffGame(game)
            else -> throw IllegalStateException("Game status must be either group or playoff.")
        }
    }

    fun findById(id: Long): Optional<Game> {
        return gameRepository.findById(id)
    }

    @PublishNews
    fun rateGame(gameId: Long, userId: Long, type: RatingType): Rating {
        val user = userRepository.findById(userId)
                .orElseThrow<IllegalArgumentException> { throw IllegalArgumentException() }
        val game = gameRepository.findById(gameId)
                .orElseThrow<IllegalArgumentException> { throw IllegalArgumentException() }
        return ratingRepository.save(Rating(type, user, game))
    }

    @PublishNews
    fun commentGame(gameId: Long, userId: Long, body: String): Comment {
        val user = userRepository.findById(userId)
                .orElseThrow<IllegalArgumentException> { throw IllegalArgumentException() }
        val game = gameRepository.findById(gameId)
                .orElseThrow<IllegalArgumentException> { throw IllegalArgumentException() }
        return commentRepository.save(Comment(body, user, game))
    }

    fun findPaginatedPlayedGames(page: Int, size: Int, sort: Sort): Page<Game> =
            gameRepository.findByHomeUserNotNullAndAwayUserNotNullAndScoreHomeNotNullAndScoreAwayNotNull(
                    PageRequest.of(page, size, sort))

    fun findGameOfUsers(page: Int, size: Int, sort: Sort, user1: User, user2: User): Page<Game> =
            gameRepository.findGameOfUsers(PageRequest.of(page, size, sort), user1, user2);

    fun findGameOfUser(page: Int, size: Int, sort: Sort, user: User): Page<Game> {
        return gameRepository.findGameOfUser(PageRequest.of(page, size, sort), user)
    }

    fun findAllOfTournament(tournament: Tournament): List<Game> =
            gameRepository.findByTournament(tournament)

    /**
     * @throws IllegalStateException There's no current tournament.
     */
    @Throws(IllegalStateException::class)
    @Transactional
    @PublishNews
    fun addTechWin(winner: User, loser: User, reporter: User): Game {
        val currentTournament = tournamentService.getCurrentTournament()
                ?: throw IllegalStateException("There's no current tournament.")
        val game = reportGame(
                winner.id!!, loser.id!!,
                ceil(getBestOfValue(currentTournament.status).value!!.toDouble() / 2).toInt(), 0)
        game.techWin = true
        game.reporter = reporter
        return game
    }

    fun placeBet(game: Game, user: User, betOnHome: Boolean): Bet {
        val bet = betRepository.findByUserAndGame(user, game)
                .map { with(it) { it.betOnHome = betOnHome; it } }
                .orElse(Bet(user = user, game = game, betOnHome = betOnHome))
        return betRepository.save(bet)
    }

    fun saveGameStats(data: String, game: Game): GameStats {
        val startedAt = JSONObject(data).getString("startedAt") // Formatting example: 2013-12-14 16:45:20 GMT
        val map = JSONObject(data).optString("map", null)
        val texture = JSONObject(data).optString("texture", null)
        val format = with (SimpleDateFormat("yyyy-MM-dd HH:mm:ss z")) {
            timeZone = TimeZone.getTimeZone("UTC")
            this
        }
        return gameStatsRepository.save(GameStats(
                data = data,
                startedAt = Timestamp(format.parse(startedAt).time),
                map = map,
                texture = texture,
                game = game
        ))
    }

    fun findMaps(page: Int, size: Int, texture: String?): Page<GameStats> {
        val pageRequest  = PageRequest.of(page, size, Sort.Direction.DESC, "created")
        return when (texture) {
            null -> gameStatsRepository.findByMapIsNotNullAndTextureIsNotNull(pageRequest)
            else -> gameStatsRepository.findByMapIsNotNullAndTextureEquals(texture, pageRequest)
        }
    }


    fun findGameStats(page: Int, size: Int): Page<GameStats> =
            gameStatsRepository.findAll(PageRequest.of(page, size, Sort.Direction.DESC, "created"))

    fun findGameStats(game: Game?): String =
            (game?.let { gameStatsRepository.findAllByGame(it) } ?: gameStatsRepository.findAll())
                    .sortedBy { it.startedAt }
                    .joinToString(prefix = "[", postfix = "]") { it.data }

    fun retrieveDistinctTextures(): List<String?> =
            gameStatsRepository.findDistinctByTextureAndMapIsNotNullAndTextureIsNotNull()

    fun findFromGameStats(game: Game?, vararg fields: String): List<Map<String, Any?>> {
        val result = mutableListOf<Map<String, Any?>>()
        val stats = JSONArray(findGameStats(game))
        for (i in 0 until stats.length()) {
            val map = mutableMapOf<String, Any?>()
            fields.forEach { field -> map[field] = stats.getJSONObject(i).opt(field) }
            result.add(map)
        }
        return result
    }

    fun updateReplayQuantity(game: Game, replayQuantity: Int): Game {
        game.replayQuantity = replayQuantity
        return gameRepository.save(game)
    }

    fun countTextures(): Map<String, Long> {
        return gameStatsRepository.findDistinctByTextureAndMapIsNotNullAndTextureIsNotNull()
                .map { texture -> (texture ?: "Unknown") to gameStatsRepository.countByTexture(texture) }
                .toMap()
    }

    fun findBetsByGame(game: Game): List<Bet> = betRepository.findAllByGame(game)

    fun findGroupGames(tournament: Tournament): List<Game> = gameRepository.findByGroupNotNullAndVoidedFalseAndTournament(tournament)

    inner class InvalidScoreException internal constructor(message: String) : RuntimeException(message)

    inner class InvalidOpponentException internal constructor(message: String) : RuntimeException(message)
}

